File: Week1_Design Patterns and Principles/Ex_1_Implementing_the_Singleton_Pattern/Code/SingletonTest.java


public class SingletonTest {

    public static void main(String[] args) {
        System.out.println("=== Testing Singleton Pattern Implementation ===\n");

        testSingletonBehavior();

        System.out.println("\n" + "=".repeat(50) + "\n");

        testLoggerFunctionality();

        System.out.println("\n" + "=".repeat(50) + "\n");

        testFromDifferentMethods();
    }

    private static void testSingletonBehavior() {
        System.out.println("Test 1: Verifying Singleton Behavior");
        System.out.println("-".repeat(40));

        Logger logger1 = Logger.getInstance();
        System.out.println("First instance: " + logger1.getInstanceDetails());

        Logger logger2 = Logger.getInstance();
        System.out.println("Second instance: " + logger2.getInstanceDetails());

        Logger logger3 = Logger.getInstance();
        System.out.println("Third instance: " + logger3.getInstanceDetails());

        if (logger1 == logger2 && logger2 == logger3) {
            System.out.println("✓ SUCCESS: All instances are the same object");
            System.out.println("✓ Singleton pattern is working correctly");
        } else {
            System.out.println("✗ FAILURE: Instances are different objects");
        }

        System.out.println("\nHash Code Verification:");
        System.out.println("logger1 hash: " + logger1.hashCode());
        System.out.println("logger2 hash: " + logger2.hashCode());
        System.out.println("logger3 hash: " + logger3.hashCode());
    }

    private static void testLoggerFunctionality() {
        System.out.println("Test 2: Testing Logger Functionality");
        System.out.println("-".repeat(40));

        Logger logger = Logger.getInstance();

        logger.log("Application started successfully");
        logger.logInfo("User login attempt");
        logger.logWarning("Memory usage is high");
        logger.logError("Database connection failed");
        logger.log("Application shutting down");
    }

    private static void testFromDifferentMethods() {
        System.out.println("Test 3: Testing from Different Methods");
        System.out.println("-".repeat(40));

        Logger loggerFromMethod1 = getLoggerFromMethod1();
        Logger loggerFromMethod2 = getLoggerFromMethod2();

        System.out.println("Logger from method 1: " + loggerFromMethod1.getInstanceDetails());
        System.out.println("Logger from method 2: " + loggerFromMethod2.getInstanceDetails());

        if (loggerFromMethod1 == loggerFromMethod2) {
            System.out.println("✓ SUCCESS: Same instance returned from different methods");
        } else {
            System.out.println("✗ FAILURE: Different instances returned from different methods");
        }

        loggerFromMethod1.log("Message from method 1 context");
        loggerFromMethod2.log("Message from method 2 context");
    }

    private static Logger getLoggerFromMethod1() {
        System.out.println("Getting logger instance from method 1...");
        return Logger.getInstance();
    }

    private static Logger getLoggerFromMethod2() {
        System.out.println("Getting logger instance from method 2...");
        return Logger.getInstance();
    }
}

File: Week1_Design Patterns and Principles/Ex_1_Implementing_the_Singleton_Pattern/Code/SingletonTest.cs
using System;
using SingletonDemo;

public static class SingletonTest
{
    public static void Run()
    {
        Console.WriteLine("\n=== Testing Singleton Pattern Implementation ===\n");

        TestSingletonBehavior();

        Console.WriteLine("\n" + new string('=', 50) + "\n");

        TestLoggerFunctionality();

        Console.WriteLine("\n" + new string('=', 50) + "\n");

        TestFromDifferentMethods();
    }

    private static void TestSingletonBehavior()
    {
        Console.WriteLine("Test 1: Verifying Singleton Behavior");
        Console.WriteLine(new string('-', 40));

        var logger1 = Logger.Instance;
        Console.WriteLine($"First instance: {logger1.InstanceDetails}");

        var logger2 = Logger.Instance;
        Console.WriteLine($"Second instance: {logger2.InstanceDetails}");

        var logger3 = Logger.Instance;
        Console.WriteLine($"Third instance: {logger3.InstanceDetails}");

        Console.WriteLine(logger1 == logger2 && logger2 == logger3
            ? "✓ SUCCESS: All instances are the same object"
            : "✗ FAILURE: Instances are different objects");

        Console.WriteLine("\nHash Code Verification:");
        Console.WriteLine($"logger1 hash: {logger1.GetHashCode()}");
        Console.WriteLine($"logger2 hash: {logger2.GetHashCode()}");
        Console.WriteLine($"logger3 hash: {logger3.GetHashCode()}");
    }

    private static void TestLoggerFunctionality()
    {
        Console.WriteLine("Test 2: Testing Logger Functionality");
        Console.WriteLine(new string('-', 40));

        var logger = Logger.Instance;

        logger.Log("Application started successfully");
        logger.LogInfo("User login attempt");
        logger.LogWarning("Memory usage is high");
        logger.LogError("Database connection failed");
        logger.Log("Application shutting down");
    }

    private static void TestFromDifferentMethods()
    {
        Console.WriteLine("Test 3: Testing from Different Methods");
        Console.WriteLine(new string('-', 40));

        var logger1 = GetLoggerFromMethod1();
        var logger2 = GetLoggerFromMethod2();

        Console.WriteLine($"Logger from method 1: {logger1.InstanceDetails}");
        Console.WriteLine($"Logger from method 2: {logger2.InstanceDetails}");

        Console.WriteLine(logger1 == logger2
            ? "✓ SUCCESS: Same instance returned from different methods"
            : "✗ FAILURE: Different instances returned from different methods");

        logger1.Log("Message from method 1 context");
        logger2.Log("Message from method 2 context");
    }

    private static Logger GetLoggerFromMethod1()
    {
        Console.WriteLine("Getting logger instance from method 1...");
        return Logger.Instance;
    }

    private static Logger GetLoggerFromMethod2()
    {
        Console.WriteLine("Getting logger instance from method 2...");
        return Logger.Instance;
    }
}
